SimTracker/SiStripDigitizer/plugins/SiStripDigitizerAlgorithm.h

```diff
@@ -170,6 +184,45 @@ class SiStripDigitizerAlgorithm {
   std::map<int, float> mapOfAPVprobabilities;
   std::map<int, std::bitset<6>> SiStripTrackerAffectedAPVMap;
   int NumberOfBxBetweenHIPandEvent;
+
+  bool includeAPVSimulation_;
+  edm::FileInPath apvBaselinesFile_tib1_;
+  edm::FileInPath apvBaselinesFile_tib2_;
+  edm::FileInPath apvBaselinesFile_tib3_;
+  edm::FileInPath apvBaselinesFile_tib4_;
+  edm::FileInPath apvBaselinesFile_tob1_;
+  edm::FileInPath apvBaselinesFile_tob2_;
+  edm::FileInPath apvBaselinesFile_tob3_;
+  edm::FileInPath apvBaselinesFile_tob4_;
+  edm::FileInPath apvBaselinesFile_tob5_;
+  edm::FileInPath apvBaselinesFile_tob6_;
+
+  unsigned int nTruePU_;
+  unsigned int apvBaselines_nBinsPerBaseline_;
+  double apvBaselines_minBaseline_;
+  double apvBaselines_maxBaseline_;
+  std::vector<double> apvBaselines_puBinEdges_;
+  std::vector<double> apvBaselines_zBinEdges_;
+
+  std::vector<std::vector<std::vector<TH1F>>> apvBaselineHistograms_tib_;
```

Comment:
I wonder if there would be any better way to organize the data than triple-vector of TH1F.

Actually, how much memory does this and the corresponding `...tob_`  consume? By quick estimate based on the text file sizes I'd guess the total to be the order of 300 kB (assuming the unnecessary double-vector members are cleaned up, and neglecting the overheads of TH1F). Is the number of z and PU bins expected to increase in the future from the current 1 (all files have a single line; btw the according to config there are 0 bins...)?

With the current approach the memory need gets multiplied by the number of EDM streams (i.e. concurrent events). One possible way to make the memory need independent of the number of EDM streams would be to provide the data to the `SiStripDigitizerAlgorithm` via EventSetup (e.g. an ESProducer that reads the data in and produces an ESProduct with the data and interface to return a properly distributed random number using a given `CLHEP::HepRandomEngine`.

Reply:
In the current text files, there are 4 bins in z, and 26 bins in PU (apologies, I did not include the correct configuration for these, I will include them).  So the memory estimate of 300kb is an underestimate.

Can you provide a similar example of how to provide these data via EventSetup?

Reply:
Already the plain `sizeof(TH1F)` gives 1000 bytes, TID and TOB together make 10, so `4*26*10*1000 = 0.99 MiB` of TH1F overhead alone (per EDM stream).

Documentation for ESProducers is here https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEventSetupHowTos
Here is an example of the creation of an ESProducer
https://github.com/cms-sw/cmssw/blob/master/RecoTracker/FinalTrackSelectors/plugins/LwtnnESProducer.cc
Here is a minimum what it required to do
https://github.com/makortel/cmssw/commit/ce7239c6fa1
I'm sure @mmusich can suggest a record to use (presumably something whose IOV is the whole job)

Reply:
If we use a single vector, with a [`PhysicsTools::Calibration::HistogramF3D`](https://github.com/cms-sw/cmssw/blob/master/CondFormats/PhysicsToolsObjects/interface/Histogram3D.h) (pu,z,V0) for each layer, the size of the value arrays (without overhead) should be about `10(layers)*4(z)*26(PU)*82(V0)*4 = 333.125kB` (we would need to implement the random sampling from a (pu,z) row of it, but from the other discussion we should do that anyway). I could write an ESProducer for this (IIRC this is also the starting point for adding it to the database, right?).
As far as I know the only MC-specific SiStrip record we have is the SiStripApvGainSimRcd, we could add this there (it's somewhat related), but can the database put two objects of different type in the same record with the same global tag? Otherwise we need to add a new record anyway.

Reply:
> we could add this there (it's somewhat related), but can the database put two objects of different type in the same record with the same global tag? 

The same record cannot hold two different C++ object abstractions. 
`SiStripApvGainSimRcd` is mapped onto `SiStripApvGain`

https://github.com/cms-sw/cmssw/blob/eb4b5d2150c7e93458cd13d13540513a99a88471/CondCore/SiStripPlugins/plugins/plugin.cc#L33

 which stores the information of the gain for each apv using four vectors, 
  * A vector<unsigned int> (v_detids) stores the detId.
  * A vector<float> (v_gains) stores the value of the gain (more than one per detId).
  * Two vector<unsigned int> (v_ibegin and v_iend) store the correspondence of the v_detids
   * and the ranges of values in v_gain. 

https://github.com/cms-sw/cmssw/blob/master/CondFormats/SiStripObjects/interface/SiStripApvGain.h#L15.

Honestly I think that abusing `SiStripApvGain` for this purpose would be (a part from complicated) ugly as hell.

Reply:
ok, thanks for clarifying - shall I add a new record then? How about `SiStripApvSimulationParametersRcd`, with the `SiStripApvSimulationParameters` C++ type?

Reply:
fine by me.

Reply:
@pieterdavid another alternative to ```std::vector<std::vector<std::vector<TH1F>>>``` is ```std::vector<std::vector<std::vector<CLHEP::RandGeneral*>>>```, which would make the random sampling more straightforward.  This appears to work, but let me know if there is a preference for using  ```PhysicsTools::Calibration::HistogramF3D```